src/dune_rules/fdo.ml

```diff
@@ -91,38 +91,39 @@ let get_profile =
   (* The dependency on the existence of profile file in source should be
      detected automatically by Memo. *)
   let f (ctx : Context.t) =
-    let path = ctx.fdo_target_exe |> Option.value_exn |> fdo_profile in
-    let profile_exists =
-      Memo.lazy_ (fun () ->
-          path |> Path.as_in_source_tree
-          |> Option.map ~f:File_tree.file_exists
-          |> Option.value ~default:false)
-    in
-    let use_profile =
-      match Mode.of_context ctx with
-      | If_exists -> Memo.Lazy.force profile_exists
-      | Always ->
-        if Memo.Lazy.force profile_exists then
-          true
-        else
-          User_error.raise
-            [ Pp.textf "%s=%s but profile file %s does not exist." Mode.var
-                (Mode.to_string Always) (Path.to_string path)
-            ]
-      | Never -> false
-    in
-    if use_profile then
-      Some path
-    else
-      None
+    Memo.Build.return
```

Comment:
This is a trivial remark, but the diff would be smaller if you push the call to `Memo.Build.return` to the leaves of the expression (`Some path` and `None`). And also easier to read IMHO due to the reduced nesting. I think there are a few other places where the same remark could apply.

Reply:
Agreed! Fixed here, and will look out for similar cases.